### PR TITLE
fix: Config with `ignores` and without `files` should not always apply

### LINKED ISSUE #17103: Bug: [new config system] `ignores` does not work without `files`
### Environment

Node version: 18.16.0
npm version: 9.6.4
Local ESLint version: 8.38.0
Global ESLint version: N/A
Operating System: Ubuntu 22.04 (WSL)

### What parser are you using?

Default (Espree)

### What did you do?

<details>
<summary>Configuration</summary>

```js
[
  {
    rules: {
      "no-unused-vars": "error",
    },
  },
  {
    ignores: ["error.js"],
    rules: {
      "no-unused-vars": "warn",
    },
  },
];
```

</details>

`./error.js` and `./warn.js`:

```js
let unusedVar;
```

Then run `eslint .`

### What did you expect to happen?

```shell
{cwd}/error.js
  1:5  error  'unusedVar' is defined but never used  no-unused-vars

{cwd}/warn.js
  1:5  warning  'unusedVar' is defined but never used  no-unused-vars
```

### What actually happened?

```shell
{cwd}/error.js
  1:5  warning  'unusedVar' is defined but never used  no-unused-vars

{cwd}/warn.js
  1:5  warning  'unusedVar' is defined but never used  no-unused-vars
```

### Link to Minimal Reproducible Example

https://github.com/filetvignon/eslint-bug-ignores

### Participation

- [X] I am willing to submit a pull request for this issue.

### Additional comments

I set 2 configuration objects: one with a rule all files and another to change that rule for all files EXCEPT `error.js`.
I omitted the `files` property in these objects; although the documentation recommends using `ignores` with `files`, [it states that this should work](https://eslint.org/docs/latest/use/configure/configuration-files-new#excluding-files-with-ignores) and omitting `files` should be the same as using `files: '**/*'`.

In reality, it works only if I explicitly set `files` in the second configuration object:

```js
[
  {
    rules: {
      "no-unused-vars": "error",
    },
  },
  {
    files: ["**/*"],
    ignores: ["error.js"],
    rules: {
      "no-unused-vars": "warn",
    },
  },
];
```

### PR DESCRIPTION
**Pull request by mdjermanovic**
#### Prerequisites checklist

- [x] I have read the [contributing guidelines](https://github.com/eslint/eslint/blob/HEAD/CONTRIBUTING.md).

#### What is the purpose of this pull request? (put an "X" next to an item)

[ ] Documentation update
[x] Bug fix ([template](https://raw.githubusercontent.com/eslint/eslint/HEAD/templates/bug-report.md))
[ ] New rule ([template](https://raw.githubusercontent.com/eslint/eslint/HEAD/templates/rule-proposal.md))
[ ] Changes an existing rule ([template](https://raw.githubusercontent.com/eslint/eslint/HEAD/templates/rule-change-proposal.md))
[ ] Add autofix to a rule
[ ] Add a CLI option
[ ] Add something to the core
[ ] Other, please explain:

Fixes #17103

#### What changes did you make? (Give an overview)

Upgraded `@humanwhocodes/config-array` dependency to include https://github.com/humanwhocodes/config-array/commit/5ed9c2c1a13afb42cd7e9d3b1b247761cb7aa040 fix, and added tests.

#### Is there anything you'd like reviewers to focus on?

~Marked as draft as it uses `@humanwhocodes/config-array` from GitHub while waiting for the release (https://github.com/eslint/eslint/issues/17103#issuecomment-1546706681)~

### CODE REVIEW

**Comment by mdjermanovic** [comment]
CI is failing on Node <= 14  due to missing `node_modules/@humanwhocodes/config-array/api.js` probably because npm <= 6 doesn't run `prepublish` scripts when installing git dependencies.

**Comment by mdjermanovic** [comment]
@nzakas it seems that the published v0.11.9 doesn't include changes from https://github.com/humanwhocodes/config-array/commit/5ed9c2c1a13afb42cd7e9d3b1b247761cb7aa040.

**Comment by mdjermanovic** [comment]
Also, per the diff between published v0.11.9 and v0.11.8, changes from https://github.com/humanwhocodes/config-array/commit/8e17f4a7378cb0b417e1103d60ef397b26d2f917 are also not present in v0.11.9.

Actually, `api.js` in v0.11.9 is the same as in v0.11.7.

**Comment by nzakas** [comment]
Okay, v0.11.10 is now pushed and should be what we need.

**Comment by mdjermanovic** [comment]
Updated to v0.11.10.

File: tests/lib/eslint/flat-eslint.js [hunk 1]
```diff
@@ -4631,6 +4631,99 @@ describe("FlatESLint", () => {
...
+
+            it("should not add json files as lint targets", async () => {
+                const engine = new FlatESLint({
+                    cwd
+                });
+
+                const results = await engine.lintFiles("foo*");
+
+                // should not lint `foo.json`
+                assert.strictEqual(results.length, 1);
```
Comment (aladdin-add) [inline tests/lib/eslint/flat-eslint.js:4720]: question: why does it have a linting error here?  To me, seems it should be no errors reported: `foo.js` is an empty file, so there are no linting errors. am i missing something?

Comment (mdjermanovic) [reply tests/lib/eslint/flat-eslint.js:4720]: It's not an error, `lintFiles` creates a [LintResult](https://eslint.org/docs/latest/integrate/nodejs-api#-lintresult-type) object for each file that was linted regardless of whether the file had errors. Errors and warnings are in `LintResult#messages`.

Comment (aladdin-add) [reply tests/lib/eslint/flat-eslint.js:4720]: I see. Misunderstood the returned value. 😄

**Review by aladdin-add — APPROVED** [review]
LGTM but would like @nzakas to take another look.

**Review by nzakas — APPROVED** [review]
LGTM. Will leave for @mdjermanovic to merge during the release.

**Pull request merged**